src/ipc/ipc4/helper.c

```diff
@@ -366,6 +369,8 @@ int ipc_comp_connect(struct ipc *ipc, ipc_pipe_comp_connect *_connect)
 		goto err;
 	}
 
+	irq_local_enable(flags);
```

Comment:
Please comment it why this is a critical section
and
> **at the moment** SOF only supports running all connected pipelines on the same core.

it will change soon enough, IPC task will always be binded to core 0, pipelines will work on all cores, what then? 
do we really want to do such temporary changes?

Reply:
Firstly we probably want that comment about why we're locking here in-line, not just as a commit message. Secondly, @marcinszkudlinski I'm not sure we need a spinlock here. It's good to prepare for the future, but we don't know yet what it will be like. I think when we'll have to run connected pipelines or components on different cores, we'll have to protect the buffer list header throughout the code, and that will be **a lot** of protecting. And then most (or all) of that code runs in threads, so we might decide to use mutexes instead of spinlocks.
Another possibility to fix this could be finding the code that crashes when the buffer isn't fully connected yet, and fixing it to ignore such half-connected buffers.

Reply:
> Another possibility to fix this could be finding the code that crashes when the buffer isn't fully connected yet, and fixing it to ignore such half-connected buffers.

I thought about this but then decided that the solution has to be generic to allow module developers to not worry about it.

Reply:
> Please comment it why this is a critical section and
> 
> > **at the moment** SOF only supports running all connected pipelines on the same core.
> 
> it will change soon enough, IPC task will always be binded to core 0, pipelines will work on all cores, what then? do we really want to do such temporary changes?

@marcinszkudlinski I think this isn't quite correct. Core 0 will manage the IPC queue, but IPC processing will be done on respective cores, just as it is done now

Reply:
> you think it would be a specific module, not some common code? Then maybe make it common... Just another option

@lyakh my original suspicion was the copier module and still is. But im not able to pinpoint exactly what in the copier module code is crashing to cause this problem. I thought it was likely the sink dev being NULL but even that doesnt help

Reply:
@lyakh @marcinszkudlinski thinking a bit more about the issue of connected pipelines being on different cores, I think this whole function ipc_comp_connect() would need some surgery to support that isnt it? This function the way it is written can only really work is both sides of the buffer are on the same core. If thats not the case, each connection would have to be handled on the core that it belongs to and then there's the question on which core to allocate the buffer itself. So, I think its too premature to talk about that at this point.